src/snowflake/connector/errors.py

```diff
@@ -5,25 +5,34 @@
 #
 
 import logging
+import os
+import traceback
 from logging import getLogger
-from typing import TYPE_CHECKING, Dict
+from typing import TYPE_CHECKING, Dict, Optional
 
 from snowflake.connector.constants import UTF8
 
 from .compat import BASE_EXCEPTION_CLASS
+from .description import CLIENT_NAME, SNOWFLAKE_CONNECTOR_VERSION
+from .secret_detector import SecretDetector
+from .telemetry import TelemetryData, TelemetryField
+from .telemetry_oob import TelemetryService
+from .time_util import get_time_millis
 
 if TYPE_CHECKING:  # pragma: no cover
     from .connection import SnowflakeConnection
     from .cursor import SnowflakeCursor
 
 logger = getLogger(__name__)
+connector_base_path = os.path.join("snowflake", "connector")
 
 
 class Error(BASE_EXCEPTION_CLASS):
     """Base Snowflake exception class."""
 
-    def __init__(self, msg=None, errno=None, sqlstate=None, sfqid=None,
-                 done_format_msg=False):
+    def __init__(self, msg: Optional[str] = None, errno: Optional[int] = None, sqlstate: Optional[str] = None,
+                 sfqid: Optional[str] = None, done_format_msg: Optional[bool] = None,
+                 connection: Optional[SnowflakeConnection] = None, cursor: Optional[SnowflakeCursor] = None):
```

Comment:
```suggestion
                 connection: Optional['SnowflakeConnection'] = None, cursor: Optional['SnowflakeCursor'] = None):
```
It's not  a problem in this file but putting these in strings can save us from some circular imports. I'd like to be consistent. If we type hint one of our own classes unless it's in the same file let's put them in strings if that makes sense to you.